omod/src/main/java/org/openmrs/module/smartonfhir/web/filter/CORSFilter.java

```diff
@@ -29,12 +29,17 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
 	        throws ServletException, IOException {
 		// CORS "pre-flight" request
 		response.addHeader("Access-Control-Allow-Credentials", "true");
```

Comment:
This is what I'd suggest for an implementation

```java
		// CORS "pre-flight" request
		if ("OPTIONS".equals(request.getMethod()) && StringUtils.isNotEmpty(request.getHeader("Access-Control-Request-Method"))) {
			response.addHeader("Access-Control-Allow-Credentials", "true");
			response.addHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE");
			response.addHeader("Access-Control-Allow-Headers", "Authorization");
			response.addHeader("Access-Control-Allow-Headers", "Content-Type");
			response.addHeader("Access-Control-Allow-Origin", "*");
			response.addHeader("Access-Control-Max-Age", "600");

			response.setStatus(HttpServletResponse.SC_OK);
			return;
		}

		filterChain.doFilter(request, response);
```

`StringUtils` in the above should be `org.apache.commons.lang3.StringUtils`.

We can eventually adopt something more sophisticated, but this should work for now.

Reply:
@ibacher after making the change, I'm getting the same old error. I'm not sure but i think some time `request.getMethod()` return value other than `OPTIONS`. 

![Screenshot from 2021-01-14 00-30-13](https://user-images.githubusercontent.com/49350053/104497230-0a6f7f00-5600-11eb-8596-738bee65c057.png)

Reply:
I checked `request.getHeader("Access-Control-Request-Method")` returns the `null` value everytime and some time `request.getMethod()` returns `GET`. I think this is the reason why `Access-Control-Allow-Origin` is missing in the header because the value of `if` condition is always false.

Reply:
Yeah, the `request.getMethod()` could theoretically be anything, but we only need to return the CORS headers for `OPTIONS` requests. Basically, the only time we need to send these headers if for the CORS [pre-flight check](https://developer.mozilla.org/en-US/docs/Glossary/preflight_request). If request.getHeader("Access-Control-Request-Method") is returning `null`, we can drop that condition (it shouldn't, but we should be pragmatic about this).

Reply:
@ibacher Thanks now things are getting more clear. I getting this error `CORS header 'Access-Control-Allow-Origin' missing` when i'm adding the `Access-Control-Allow-Origin` header only for OPTIONS, but when i put it outside the if body, everything works fine. I'm not sure but i think other methods also require `Access-Control-Allow-Origin` header ([CORS Missing origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSMissingAllowOrigin)). working code for me:
```
response.addHeader("Access-Control-Allow-Origin", "*");

		if ("OPTIONS".equals(request.getMethod())) {
			response.addHeader("Access-Control-Allow-Credentials", "true");
			response.addHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE");
			response.addHeader("Access-Control-Allow-Headers", "Authorization");
			response.addHeader("Access-Control-Allow-Headers", "Content-Type");
			response.addHeader("Access-Control-Max-Age", "600");
			
			response.setStatus(HttpServletResponse.SC_OK);
			return;
		}

filterChain.doFilter(request, response);
```